Subscribe and unsubscribe push notifications on sign-in and sign-out

arosiclair opened the pull request:
### Details
These changes fix issues with push notifications being unsubscribed when putting the app in the background using the back button.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/244304   

### Tests
1. Sign into NewDot on Android
2. Send a message to your user
3. Verify a push notification displays for the message after a moment
4. Use the back button (or swipe from edge if gestures are enabled) to put the app in the background
5. Send another message to your user
6. Verify a push notification displays for the message after a moment
7. Sign out
8. Send another message to your user
9. Verify a notification is not received

- [ ] Verify that no errors appear in the JS console

### Offline tests
None

### QA Steps
Same as Tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>
None

</details>

<details>
<summary>Mobile Web - Chrome</summary>

None

</details>

<details>
<summary>Mobile Web - Safari</summary>

None

</details>

<details>
<summary>Desktop</summary>

None

</details>

<details>
<summary>iOS</summary>

![IMG_07D7A70335CE-1](https://user-images.githubusercontent.com/6833644/209711645-f67255b4-f419-4752-99ef-ef6c31f4e6e5.jpeg)

</details>

<details>
<summary>Android</summary>

![Screenshot_20221227-125439](https://user-images.githubusercontent.com/6833644/209709840-a1b1089a-1004-485e-9121-ee45bb337b13.png)

</details>

arosiclair committed:
move session clean up logic to the logOut action

fixes issues with push notifications getting unsubscribed when putting the app in the background using the back button on Android

arosiclair committed:
move push notification registration to auth screens

arosiclair committed:
use early return

marcaaron commented on src/Expensify.js:
```diff
@@ -115,23 +115,18 @@ class Expensify extends PureComponent {
-        if (currentAccountID && (currentAccountID !== previousAccountID)) {
-            PushNotification.register(currentAccountID);
+    componentDidUpdate() {
+        if (!(this.state.isNavigationReady || this.state.isSplashShown)) {
```
```suggestion
        if (!this.state.isNavigationReady || !this.state.isSplashShown) {
```

Inverting a statement like this always confuses me so correct me if I have this wrong... but I _think_ we want to proceed only when **both** of these are `true`? If so, we can return early if either one is `false` and it is the same as saying both must be `true`?

fwiw, I usually have to write the possibilities out to see if I have it right - in this case I factored that if one of these values is true and one is false then the inner conditional evaluates to `true` which when inverted would be `false` and the code will run (which I think is not the correct behavior).

arosiclair replied:
Yeah taking a closer look at this you're right. My inversion was a bit off here

Julesssss replied:
I think this has been resolved?

marcaaron commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -132,7 +133,6 @@ class AuthScreens extends React.Component {
         if (this.unsubscribeGroupShortcut) {
             this.unsubscribeGroupShortcut();
         }
-        Session.cleanupSession();
```
Not following this change. Why should we only do these tasks when the user explicitly signs out? 

It seems like on sign out this series of events will happen:

- `authToken` will clear
- `AuthScreens` will unmount as there is no `authToken`
- Cleanup happens as a side effect of losing the `authToken`

arosiclair replied:
We're moving this out of `componentWillUnmount` because that also runs when you press the back button to leave the app on Android. Are there other flows where we should clear session? Like when the authToken expires?

arosiclair replied:
Just tried this [SO](https://stackoverflowteams.com/c/expensify/questions/7826), which gets the token to expire however NewDot doesn't really react (on this branch and `main`).

marcaaron replied:
Oh darn, to clarify, you're saying that the `AuthScreens` `componentWillUnmount()` runs when the back button is pressed to leave the app on Android?

I wouldn't mind looking at this for a second. Why does that happen? And does it happen on a release build or just dev? Does `componentDidMount()` get called again when we bring the app to the foreground?

My working understanding of stuff is that we can use the `AuthScreens` mounting and unmount lifecycles as a way to "set up" or "tear down" things that should always happen when the user logs in or out. That is 100% based on the assumption that `componentWillUnmount()` only fires when we previously had an `authToken` but lost it for some reason (manually signing out or manually clearing storage are the two cases I can think of).

I think this is kind of easier to reason about because it is very binary - you either have an `authToken` or you don't. And if you had one before, but don't have one now then that is a "user logged out" scenario.

`AuthScreens` unmounting for some other reason definitely creates a problem for that idea 😄 

>Are there other flows where we should clear session? Like when the authToken expires?

The `authToken` can expire, but under normal conditions it will get refreshed. A user could manually clear their `authToken` from storage and that would effectively be the same as signing out without explicitly calling `signOut()`.

arosiclair replied:
> Oh darn, to clarify, you're saying that the AuthScreens componentWillUnmount() runs when the back button is pressed to leave the app on Android?

Yup that's correct. Specifically when you're on the home screen, press back to leave the app. This kills the app and AuthScreens gets unmounted.

> I wouldn't mind looking at this for a second. Why does that happen?

I haven't looked too closely, but I think it's a quirk with `react-navigation`. Pressing the back button usually pops a screen off the stack so I'm guessing that also happens even when you press back to exit the app. Unmounting screens might also just be part of the teardown process when the app gets killed.

>  And does it happen on a release build or just dev?

Both. This issue has been breaking notifications for me (and many others probably) on the prod app. Eventually I was able repro the issue dev.

marcaaron replied:
Ok cool I think in that case we need to hook the "clean up" parts of the app an Onyx handler that detects that the `authToken` has been cleared or something? Otherwise, it seems there is code that won't run until we explicitly call `signOut()` and seems like the only place where we do that is here:

https://github.com/Expensify/App/blob/0fcebd329567040d8832a314205ab61e602488fb/src/pages/settings/InitialSettingsPage.js#L214-L222

marcaaron replied:
I think it would also be good to maybe mention somewhere in a code comment about the quirky android behavior. I feel like the code is in a slightly inconsistent state where:

- "setup" tasks mostly are find to happen when `AuthScreens` mounts
- "teardown" tasks should not happen when `AuthScreens` unmounts

And I think what we've really got is something like...

- things that need to happen when AuthScreens mounts/unmounts (e.g. subscribing to Pusher, calling OpenApp)
- things that need to happen when a user is signed in or signed out (e.g. de/registering for Push Notifications)

arosiclair replied:
Yeah I think using an Onyx handler could be a good alternative. We could use that specifically for push notifications subscribe/unsubscribe and leave the other session cleanup as it was. I'll draft that soon

arosiclair committed:
fix check for navigation and splash state

Co-authored-by: Marc Glasser <marc.aaron.glasser@gmail.com>

marcaaron commented:
Looks like there is a conflict (left a couple more notes in the thread above).

arosiclair committed:
Merge branch 'main' into arosiclair-android-push-unsubscribe-fix

arosiclair committed:
re-add Session.cleanupSession and use an Onyx connection to manage push notification subscription

arosiclair committed:
added doc explaining Android quirks

arosiclair committed:
added unit test for push subscription

arosiclair committed:
set session to null to match our optimistic LogOut behavior

arosiclair commented:
Alright this is ready for another review. I've moved subscribe/unsubscribe logic to an Onyx.connect handler and also added a unit test to make sure `register()` and `deregister()` are called when our session changes in Onyx.

marcaaron commented on src/libs/actions/Session/index.js:
```diff
@@ -23,6 +23,32 @@ Onyx.connect({
+/**
+ * Manage push notification subscriptions on sign-in/sign-out.
+ *
+ * We would normally do this when AuthScreens mounts and unmounts. However on Android,
+ * AuthScreens unmounts when the app is closed with the back button so we manage the
+ * push subscription when the session changes here.
```
```suggestion
 * On Android, AuthScreens unmounts when the app is closed with the back button so we manage the
 * push subscription when the session changes here.
```

I think this is a good enough warning and feels like we are standing by this solution a bit more. After working through the problem it sounds really undesirable to use `AuthScreens` for this.

marcaaron commented on src/libs/actions/Session/index.js:
```diff
@@ -23,6 +23,32 @@ Onyx.connect({
+    callback: (session) => {
+        const accountID = lodashGet(session, 'accountID');
+
+        if (accountID && previousAccountID !== accountID) {
```
So if we have an `accountID` and the `previousAccountID` doesn't equal it then we register a new account.

Are we assuming the `previousAccountID === ''` here or do we not care about what value it was before?

marcaaron replied:
I guess I'm wondering if there is a case where the `accountID` can have a value but be different from `previousAccountID`. I'm not sure if that's possible though - maybe it always has to be at least `null`.

arosiclair replied:
Yeah as long as the value changed, we would want to register the new account ID. This should replace the old registration if there was one.

marcaaron commented on src/libs/actions/Session/index.js:
```diff
@@ -23,6 +23,32 @@ Onyx.connect({
+        if (!accountID && previousAccountID !== accountID) {
+            PushNotification.deregister();
+            PushNotification.clearNotifications();
+            previousAccountID = accountID;
```
Is there a case where we would not want to set the `previousAccountID`? Or could we just do that once at the bottom of this function and always update it to the latest?

arosiclair replied:
I dunno if I like the "current" and "next" naming since the accountID has already changed by the time this callback runs but I do think this logic would work too

marcaaron commented on src/libs/actions/Session/index.js:
```diff
@@ -23,6 +23,32 @@ Onyx.connect({
+ * AuthScreens unmounts when the app is closed with the back button so we manage the
+ * push subscription when the session changes here.
+ */
+let previousAccountID = '';
```
I think accountID are number?

marcaaron replied:
Maybe can make it `undefined` initially? 

```suggestion
let previousAccountID;
```

arosiclair replied:
> I think accountID are number?

Yeah that's actually right it looked like a string in logs

marcaaron commented on src/libs/actions/Session/index.js:
```diff
@@ -446,13 +470,13 @@ export {
     updatePasswordAndSignin,
     signIn,
     signInWithShortLivedAuthToken,
+    cleanupSession,
```
NAB but this does not really need to be moved.

arosiclair replied:
Just a bad revert job from me 😅

marcaaron commented on tests/unit/PushNotificationsTest.js:
```diff
@@ -0,0 +1,47 @@
+
+// The Onyx handler for subscribing and unsubscribing is in this module
+// so we must import it
+import '../../src/libs/actions/Session';
```
Thanks for writing tests 🙇 This test is great - but not sure if it's _really_ a unit test for the `PushNotifications` lib. We are testing the code in the actions file.

**non-blocking:** I would suggest incorporating this test into the existing `SessionTest.js` file. We have a `TestHelper.signInWithTestUser()` method that will simulate signing in the user and calls the action which signs in the user.

Probably we could:

- Figure out some way to call `Session.signOut()`
- Sign in a user with `signInWithTestUser()` (check the mock like we are doing in this test)
- Sign out a user (check the mock again)

marcaaron commented on tests/unit/PushNotificationsTest.js:
```diff
@@ -0,0 +1,47 @@
+
+        test('Push notifications are subscribed after signing in', () => {
+            Onyx.merge(ONYXKEYS.SESSION, {
+                accountID: 'test-id',
```
would use the correct type here. Even if this is just `1`.

arosiclair committed:
use undefined initial value

Co-authored-by: Marc Glasser <marc.aaron.glasser@gmail.com>

arosiclair committed:
simpler explanation

Co-authored-by: Marc Glasser <marc.aaron.glasser@gmail.com>

arosiclair committed:
move push notification subscription test to SessionTest

arosiclair committed:
simplify previousAccountID updates

arosiclair commented:
This is ready for another review however I'm gonna put this on hold for https://github.com/Expensify/App/pull/13844 which is going to change subscription from using `accountID` to using `pushNotificationID`. The pattern here should be more or less the same, however the timing of subscribing/unsubscribing will change since the `pushNotificationID` will be sent in `OpenApp` and `ReconnectApp`.

Julesssss commented on src/libs/actions/Session/index.js:
```diff
@@ -23,6 +23,31 @@ Onyx.connect({
+    callback: (session) => {
+        const accountID = lodashGet(session, 'accountID');
+
+        if (accountID && previousAccountID !== accountID) {
```
I wonder if we early return if `previousAccountID !== accountID` is false, then use an if/else condition for the register/deregister logic.

Julesssss replied:
```
if (previousAccountID == accountID) {
    return; 
}

if (accountID) {
    PushNotification.register(accountID);
} else {
    PushNotification.deregister();
    PushNotification.clearNotifications();
}
```

marcaaron replied:
+1 I like this suggestion

Julesssss reviewed (approved):
I know this is still held, but I verified that we remain authenticated when closing the app via the back button 👍 

![rn_image_picker_lib_temp_79b164ae-f4cb-4904-88ae-66884b8e71d3](https://user-images.githubusercontent.com/10736861/211376564-d3c93743-2dc8-466d-b20c-dfb62da45198.png)

arosiclair commented:
Taking this off hold since we have a hold on the other PR now

Julesssss commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

Julesssss commented:
@arosiclair should we add tests for web/Desktop, to ensure they still receive notifications? I'm not seeing them on your branch, but maybe this is expected?

Julesssss commented:
I ran out of time and still need to verify iOS on a physical device. Will continue tomorrow.

arosiclair commented:
> @arosiclair should we add tests for web/Desktop, to ensure they still receive notifications? I'm not seeing them on your branch, but maybe this is expected?

In dev I couldn't get web notifications on `localhost` but I was able to on `127.0.0.1`.

Julesssss reviewed (dismissed):
I finished testing iOS this morning, and it tested well. There's just the one outstanding comment remaining now.

arosiclair committed:
simplify subscribe/unsubscribe logic

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3902620334) to staging by @Julesssss in version: 1.2.53-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

Julesssss commented:
![Screenshot_20230112-151233](https://user-images.githubusercontent.com/10736861/212123138-40bdb2a4-cad6-4375-bdac-b6cc5fe87b56.png)

Verified on staging 👍

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3910716992) to production by @Julesssss in version: 1.2.53-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
